File: codes/leetcode/DesignUndergroundSystem.cpp
class UndergroundSystem {
private:
    unordered_map<int, pair<string, int>> log;
    unordered_map<string, pair<int, int>> history;
    
public:
    UndergroundSystem() {
        
    }
    
    void checkIn(int id, string station, int time) {
        log[id] = {station, time};
    }
    
    void checkOut(int id, string station, int time) {
        pair<string, int> entry = log[id];
        
        string startStation = entry.first;
        int startTime = entry.second;
        
        string trip = startStation + station;
        
        auto &data = history[trip];        
        auto &tripCount = data.first;
        auto &totalTime = data.second;
        
        tripCount++;
        totalTime += time-startTime;
    }
    
    double getAverageTime(string startStation, string endStation) {
        string trip = startStation + endStation;
        
        pair<int, int> data = history[trip];
        int tripCount = data.first;
        int totalTime = data.second
            ;
        return totalTime/(1.00*tripCount);
    }
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem* obj = new UndergroundSystem();
 * obj->checkIn(id,stationName,t);
 * obj->checkOut(id,stationName,t);
 * double param_3 = obj->getAverageTime(startStation,endStation);
 */

File: codes/leetcode/DesignUndergroundSystem.py
# ************************************************
# username  :   smmehrab
# fullname  :   s.m.mehrabul islam
# email     :   smmehrabul-2017614964@cs.du.ac.bd
# institute :   university of dhaka, bangladesh
# reg       :   2017614964
# ************************************************

class UndergroundSystem:
    def __init__(self):
        self.history = {}
        self.current = {}

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.current[id] = (stationName, t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        start_station, start_time = self.current.pop(id)
        trip = (start_station, stationName)
        time = t - start_time
        if trip in self.history:
            total_time, count = self.history[trip]
            self.history[trip] = (total_time + time, count + 1)
        else:
            self.history[trip] = (time, 1)

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        trip = (startStation, endStation)
        total_time, count = self.history[trip]
        return total_time / count

# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)